### PR TITLE
Seed management to improve reproducibility

### LINKED ISSUE #1321: Ensure TestWorld seed is propagated to all RNGs
I just realized that we probably want to use the same seed for rng and test world (it is supported via `TestWorldConfig` struct). That way we can make it reproducible if it ever fails.

This was originally raised in the context of tests that use `rng.gen()` to generate random input values, but we should consider all kinds of randomness, e.g., proptests.

_Originally posted by @akoshelev in https://github.com/private-attribution/ipa/pull/1316#discussion_r1777874860_

### PR DESCRIPTION
**Pull request by andyleiserson**
Changes to tests and test infrastructure to be able to reproduce test runs from a seed.

This just updates one example of each kind of test I could think of to demonstrate how to achieve reproducibility. There are still many other tests that would need to be updated to match.

Fixes #1321

### CODE REVIEW

File: ipa-core/src/protocol/ipa_prf/aggregation/mod.rs [hunk 1]
```diff
@@ -503,37 +501,38 @@ pub mod tests {
...
-                let row: [u32; PROP_BUCKETS] = array::from_fn(|_| rng.gen_range(0..1 << tv_bits));
+            for row in &inputs {
                 for (exp, val) in expected.iter_mut().zip(row) {
                     *exp = min(*exp + val, (1 << PropHistogramValue::BITS) - 1);
                 }
-                row
-            })
-            .take(len)
-            .collect();
+            }
```
Comment (andyleiserson) [inline ipa-core/src/protocol/ipa_prf/aggregation/mod.rs:516]: These changes are not essential to this PR, but it seemed clearer not to have two different seeds, and using proptest for input generation enables using its other capabilities like shrinking.

File: ipa-core/src/lib.rs [hunk 2]
```diff
@@ -277,15 +279,31 @@ pub(crate) mod test_executor {
...
     use std::future::Future;
 
+    use rand::{rngs::StdRng, thread_rng, Rng, SeedableRng};
+
+    // These routines use `FnOnce` because it is easier than dealing with lifetimes of
+    // `&mut rng` borrows in futures. If there were a need to support multiple
+    // iterations (or to make the API use `Fn` to match the shuttle version), the
+    // simplest strategy might be to seed per-iteration RNGs from a primary RNG, like
+    // `TestWorld::rng`.
+
```
Comment (akoshelev) [inline ipa-core/src/lib.rs:303]:
```suggestion
```

File: ipa-core/src/test_fixture/world.rs [hunk 3]
```diff
@@ -190,6 +200,13 @@ impl Default for TestWorld {
     }
 }
 
+impl TestWorld {
+    #[must_use]
+    pub fn with_seed(seed: u64) -> Self {
```
Comment (akoshelev) [inline ipa-core/src/test_fixture/world.rs:205]: Default values for generic parameters is a blessing and a curse.

`impl TestWorld<NotSharded>` that exists on the line 266 is probably a better place for this fn.

**Review by akoshelev — APPROVED** [review]

**Comment by akoshelev** [comment]
Thank you for making this change @andyleiserson, we are getting closer to fully reproducible tests with it

File: ipa-core/src/test_fixture/world.rs (hunk 3 above)
Comment (andyleiserson) [reply ipa-core/src/test_fixture/world.rs:205]: Is there a reason to implement `Default` only for `TestWorld<NotSharded>`?

Comment (akoshelev) [reply ipa-core/src/test_fixture/world.rs]: backward compatibility was the only one - I didn't want to change a bunch of non-sharded tests when introducing the in-memory sharding infrastructure

**Pull request merged**